Fix part of #9749: Migration of improvements-tab, statistics-tab, feedback-tab in Angular

heyimShivam opened the pull request:
## Overview
AngularJs to Angular

1. This PR fixes or fixes part of #9749.
2. This PR does the following: AngularJS to Angular

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

https://user-images.githubusercontent.com/77006897/190421099-5916f384-8ee2-4277-87e4-3ca8e257eb82.mp4




#### Proof of changes on desktop with slow/throttled network

https://user-images.githubusercontent.com/77006897/190421099-5916f384-8ee2-4277-87e4-3ca8e257eb82.mp4



#### Proof of changes on mobile phone

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

heyimShivam committed:
small PR nits

heyimShivam commented:
@JeeveshGarg PTAL to the changes of `scripts/typescript_checks.py file`

heyimShivam committed:
split PR done

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into smallPR1

heyimShivam committed:
done

heyimShivam committed:
done

heyimShivam committed:
done

heyimShivam commented:
@ashutoshc8101 PTAL

ashutoshc8101 commented on core/templates/components/common-layout-directives/common-elements/answer-content-modal.component.ts:
```diff
@@ -0,0 +1,45 @@
+// limitations under the License.
+
+/**
+ * @fileoverview component for AnswerContentModal.
```
```suggestion
 * @fileoverview Component for answer content modal.
```

heyimShivam replied:
done

heyimShivam committed:
some

ashutoshc8101 commented on core/templates/components/shared-component.module.ts:
```diff
@@ -326,7 +331,12 @@ import { NgbModalModule } from '@ng-bootstrap/ng-bootstrap';
     QuestionSuggestionEditorModalComponent,
     QuestionSuggestionReviewModalComponent,
-    ReviewTestPageComponent
+    ReviewTestPageComponent,
+    AnswerContentModalComponent,
+    VisualizationSortedTilesComponent,
+    OppiaVisualizationClickHexbinsComponent,
+    OppiaVisualizationEnumeratedFrequencyTableComponent,
```
Sort these here and elsewhere

heyimShivam replied:
done

ashutoshc8101 commented on core/templates/components/common-layout-directives/common-elements/answer-content-modal.component.ts:
```diff
@@ -0,0 +1,45 @@
+// limitations under the License.
+
+/**
+ * @fileoverview component for AnswerContentModal.
```
```suggestion
 * @fileoverview Component for answer content modal.
```

heyimShivam replied:
done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.ts:
```diff
@@ -16,245 +16,254 @@
+  threadIsStale: boolean;
+  threadData: FeedbackThread[];
+  messageSendingInProgress: boolean;
+  tmpMessage: {
```
```suggestion
  tmpMessage: {
```
This variable name be more declarative.

heyimShivam replied:
updated

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/statistics-tab/statistics-tab.component.ts:
```diff
@@ -17,132 +17,163 @@
-        );
-      };
+  constructor(
+     private alertsService: AlertsService,
+     private computeGraphService: ComputeGraphService,
+     private explorationDataService: ExplorationDataService,
+     private explorationStatsService: ExplorationStatsService,
+     private readOnlyExplorationBackendApiService:
+       ReadOnlyExplorationBackendApiService,
+     private routerService: RouterService,
+     private stateInteractionStatsService: StateInteractionStatsService,
+     private statesObjectFactory: StatesObjectFactory,
+     private urlInterpolationService: UrlInterpolationService,
```
Sort these here and elsewhere

heyimShivam replied:
done

heyimShivam committed:
updated

heyimShivam committed:
done

heyimShivam commented:
@ashutoshc8101 PTAL

heyimShivam committed:
type error

heyimShivam committed:
component

U8NWXD reviewed (approved):
LGTM. I think adding the files to `typescript_checks.py` is okay since you're just changing the names of existing files and adding tests

JeeveshGarg commented:
> @JeeveshGarg PTAL to the changes of `scripts/typescript_checks.py file`

LGTM!

nithinrdy commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -153,25 +153,25 @@ <h2>Exploration Feedback</h2>
     font-size: 19px;
   }
-  feedback-tab .oppia-thread-button {
+  .oppia-thread-button {
```
Please make sure the CSS rules are scoped, so that they only apply to this component.

heyimShivam replied:
This css property is only applicable to this component only, as we are using this className only here.

nithinrdy replied:
Yep, but I was thinking it's possible a class with the same name(s) may be added in the future (for example, the class `oppia-icon-container` on line 170 is a very generic name and could be used in a different place in the future), and that may cause issues.

heyimShivam replied:
updated

nithinrdy commented on core/templates/pages/exploration-editor-page/suggestion-modal-for-editor-view/exploration-editor-suggestion-modal.component.ts:
```diff
@@ -0,0 +1,120 @@
+  @Input() suggestionIsHandled: boolean;
+  @Input() suggestionIsValid: boolean;
+  @Input() suggestionStatus: string;
+  @Input() threadUibModalInstance: NgbActiveModal;
```
Isn't this (and in other places too) NgbModal now i.e. `threadNgbModalInstance`?

heyimShivam replied:
replaced

nithinrdy reviewed (changes requested):
Thanks! Left a few comments.

heyimShivam committed:
suggested changes

heyimShivam committed:
cMerge remote-tracking branch 'upstream/develop' into smallPR1

heyimShivam commented:
@nithinrdy, @kevintab95  PTAL

heyimShivam commented:
@gp201 PTAL

nithinrdy commented:
@heyimShivam PTAL, replied to your comment.

heyimShivam committed:
done

heyimShivam commented:
@nithinrdy PTAL

gp201 reviewed (approved):
LGTM

heyimShivam committed:
Merge branch 'develop' into smallPR1

heyimShivam commented:
Please do not merge this PR even if the requirements are met; I want to re-test this PR again to makesure everything goes fine after the merge.

kevintab95 commented on extensions/visualizations/oppia-visualization-sorted-tiles.component.html:
```diff
@@ -0,0 +1,36 @@
+<div *ngIf="options.header">
+  <strong>{{ options.header }}</strong>
+</div>
+<!-- | orderBy: -datum.frequency; -->
```
Please avoid commented code.

heyimShivam replied:
removed

kevintab95 commented on extensions/visualizations/oppia-visualization-frequency-table.directive.spec.ts:
```diff
@@ -0,0 +1,59 @@
+        OppiaVisualizationFrequencyTableComponent
+      ],
+      providers: [
+      ],
```
Should we remove this property if it is empty?

heyimShivam replied:
removed

kevintab95 commented on extensions/visualizations/oppia-visualization-enumerated-frequency-table.directive.spec.ts:
```diff
@@ -16,74 +16,42 @@
+      OppiaVisualizationEnumeratedFrequencyTableComponent);
+    component = fixture.componentInstance;
 
-  it('should display first answer and hide the second answer', () => {
```
Should there be a replacement for this unit test?

heyimShivam replied:
this test case is not required as the .ts file is completely covered by other test cases.

I don't have idea how to test HTML code in jasmine, can you please help?

kevintab95 replied:
> this test case is not required as the .ts file is completely covered by other test cases.

Hm, that is odd, could you explain how migrating the file resulted in this? And which test covers this?


Regarding DOM testing, perhaps this helps: https://angular.io/guide/testing-components-basics#nativeelement

heyimShivam replied:
test added

kevintab95 commented on extensions/visualizations/oppia-visualization-enumerated-frequency-table.directive.spec.ts:
```diff
@@ -16,74 +16,42 @@
+    component.data = [1, 0, 0];
   });
 
-  it('should display second answer and hide the first answer', () => {
```
Ditto: Missing unit test

heyimShivam replied:
same as above

kevintab95 replied:
Which test covers this?

heyimShivam replied:
test added

kevintab95 commented on extensions/visualizations/oppia-visualization-click-hexbins.directive.spec.ts:
```diff
@@ -16,100 +16,89 @@
-    $rootScope.$digest();
-  });
 
-  it('should group the two answers as two distinct hexagons', () => {
```
Why are these tests removed? Ditto below.

heyimShivam replied:
same as above

kevintab95 replied:
Which specific tests cover this scenario?

heyimShivam replied:
test added

kevintab95 commented on extensions/visualizations/oppia-visualization-click-hexbins.directive.html:
```diff
@@ -1,32 +1,33 @@
+      <path *ngFor="let hexbin of hexbins; index as index"
+            [ngClass]="['click-hexbin-hexagon', 'e2e-test-hexagon-' + index]"
+            [attr.fill]="getFillColor(hexbin)"
+            [attr.d]="'M {{hexbin.x, {{hexbin.y}}{{hexagon}}'"
```
Please double check if the "d" attribute value works the same as before.

heyimShivam replied:
Kevin, i am unable to test the components starting with `oppia-visualization-`. can you please share the steps to test this component?

heyimShivam replied:
can you please share the steps to test this component?

kevintab95 replied:
I believe this is used in the improvements tabs in the exploration editor page. But I think this tab is currently disabled and not in use. Though @vojtechjelinek can confirm this.

That said, it should be possible to test functionality of this directive / component indepedently of where it is used. I wonder if we can render this on a different page for testing purposes on the local machine?

vojtechjelinek replied:
Yes I think it is not in use, it should still be possible to test it in some way.

heyimShivam replied:
i think testing is not possible, due to too many console errors as components are dependent on other components.

seanlip replied:
Is it possible to just enable the improvements tab temporarily so that you don't run into import errors?

kevintab95 replied:
Please note that the attr here is actually different from the previous code.

kevintab95 replied:
> i think testing is not possible, due to too many console errors as components are dependent on other components.

Would it be possible to pass in the relevant values to the child component to test if the functionality works -- I understand some of the backend parts for this functionality is disabled / missing.

kevintab95 commented on core/templates/pages/exploration-editor-page/suggestion-modal-for-editor-view/exploration-editor-suggestion-modal.component.html:
```diff
@@ -0,0 +1,58 @@
+<div class="modal-header">
```
How was this modal tested?

kevintab95 replied:
Has this been addressed?

heyimShivam replied:
we can not test this component, as in feedback tab we are passing false value to *ngIf statement `*ngIf="activeThread.isSuggestionThread() && index == 0"` && `isSuggestionThread(): boolean {
    return false;
  }`

seanlip replied:
@heyimShivam are you saying this is dead code? If so, should this be deleted?

heyimShivam replied:
I think this component is made for future use.

seanlip replied:
No, it is not -- we used to have suggestions, but then got rid of them because they weren't useful.

heyimShivam replied:
removed

kevintab95 commented on core/templates/pages/exploration-editor-page/statistics-tab/charts/pie-chart.component.spec.ts:
```diff
@@ -16,187 +16,128 @@
+import { FormsModule } from '@angular/forms';
 import { of } from 'rxjs';
-
-describe('Pie Chart component', function() {
```
Why are several tests missing?

heyimShivam replied:
the deleted test case of these files is got covered in other test cases.

kevintab95 replied:
Can you explain which specific test(s) cover the deleted ones? This is important to know before me decide to delete tests altogether.

heyimShivam replied:
test added back `earlier: should initialize afterview init' covering the test cases`

kevintab95 commented on core/templates/pages/exploration-editor-page/improvements-tab/needs-guiding-responses-task.component.ts:
```diff
@@ -13,25 +13,48 @@
+   }
+
+   ngOnInit(): void {
+     this.sortedTilesData = [...this.stats.answerStats];
```
Why has the assignment here changed compared to the original?

heyimShivam replied:
answerStats is of type read-only, I think in angularJS most of the type error remains undetected

kevintab95 commented on core/templates/components/common-layout-directives/common-elements/answer-content-modal.component.spec.ts:
```diff
@@ -0,0 +1,74 @@
+    fixture.detectChanges();
+  });
+
+  it('should evalute scope letiables values correctly', fakeAsync(() => {
```
What are "letiables"?

heyimShivam replied:
it was there in the angularJS code, `has been updated now`

kevintab95 commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.ts:
```diff
@@ -16,245 +16,254 @@
+    private userService: UserService,
+  ) { }
+
+  _resetTmpMessageFields(): void {
```
It looks like we have replaced "tmpMessage" with the "feedbackMessage" as part of these changes, should we also update the method names to so that they are in sync with the change?

heyimShivam replied:
updated

kevintab95 commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -91,49 +93,47 @@ <h2>Exploration Feedback</h2>
-                    id="$ctrl.tmpMessageText" rows="6"
-                    ng-disabled="$ctrl.messageSendingInProgress" focus-on="tmpMessageText">
+          <textarea class="form-control e2e-test-feedback-response-textarea oppia-autofocus" [(ngModel)]="feedbackMessage.text"
+                    id="tmpMessageText" rows="6"
+                    [disabled]="messageSendingInProgress" (focus)="tmpMessageText">
```
Should we update the occurences of "tmpMessageText" to match the new variable feedbackMessage?

heyimShivam replied:
done

kevintab95 reviewed (changes requested):
Thanks @heyimShivam! Left some comments. PTAL

heyimShivam committed:
upstream cnflicts

heyimShivam committed:
Merge remote-tracking branch 'origin/smallPR1' into smallPR1

heyimShivam committed:
done

heyimShivam commented:
@kevintab95 PTAL

heyimShivam commented:
Kevin, i am unable to test the components starting with oppia-visualization-. can you please share the steps to test this component?

kevintab95 reviewed (changes requested):
Thanks @heyimShivam! Left some comments. PTAL

kevintab95 commented:
@vojtechjelinek PTAL https://github.com/oppia/oppia/pull/16060#discussion_r979450291 when you get a chance. Thanks!

vojtechjelinek commented on core/templates/pages/exploration-editor-page/improvements-tab/improvements-tab.component.ts:
```diff
@@ -13,109 +13,159 @@
+import { ExplorationImprovementsTaskRegistryService, StateTasks } from 'services/exploration-improvements-task-registry.service';
+import { RouterService } from '../services/router.service';
+
+type HbrTask = HighBounceRateTask;
+type IflTask = IneffectiveFeedbackLoopTask;
+type NgrTask = NeedsGuidingResponsesTask;
+type SiaTask = SuccessiveIncorrectAnswersTask;
```
Why is this needed? IMO it just makes the code more confusing.

heyimShivam replied:
removed

vojtechjelinek commented on core/templates/pages/exploration-editor-page/statistics-tab/charts/pie-chart.component.ts:
```diff
@@ -16,74 +16,101 @@
-      });
+   directiveSubscriptions = new Subscription();
+   // TODO(#13015): Remove use of unknown as a type.
+   chart: unknown;
```
Why you cannot type this now?

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/statistics-tab/statistics-tab.component.ts:
```diff
@@ -17,132 +17,163 @@
+  selector: 'oppia-statistics-tab',
+  templateUrl: './statistics-tab.component.html'
+})
+export class StatisticsTabComponent implements
+   OnInit, OnDestroy {
```
```suggestion
export class StatisticsTabComponent implements
    OnInit, OnDestroy {
```

heyimShivam replied:
done

vojtechjelinek commented on extensions/visualizations/oppia-visualization-sorted-tiles.component.ts:
```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
```suggestion
// Copyright 2022 The Oppia Authors. All Rights Reserved.
```
ditto elsewhere

heyimShivam replied:
done

vojtechjelinek commented on extensions/visualizations/oppia-visualization-sorted-tiles.component.ts:
```diff
@@ -0,0 +1,90 @@
+
+  isAnswerTooLong(index: number): boolean {
+    return this.utilsService.isOverflowing(
+      $('.oppia-visualization-sorted-tile-answer.answer-' + index).get(0)
```
Replace the usage of jQuery.

heyimShivam replied:
updated PTAL

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

heyimShivam committed:
updated with upstream

heyimShivam committed:
changes fixed

heyimShivam commented:
@vojtechjelinek PTAL

heyimShivam committed:
done

heyimShivam committed:
updated

vojtechjelinek commented on core/templates/components/common-layout-directives/common-elements/answer-content-modal.component.spec.ts:
```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
```suggestion
// Copyright 2022 The Oppia Authors. All Rights Reserved.
```

vojtechjelinek replied:
ditto elsewhere but this is a nit

vojtechjelinek commented on core/templates/pages/exploration-editor-page/exploration-editor-page.component.html:
```diff
@@ -122,20 +122,20 @@
 
-    <improvements-tab ng-if="$ctrl.isImprovementsTabEnabled()" ng-show="$ctrl.getActiveTabName() === 'improvements'">
-    </improvements-tab>
+    <oppia-improvements-tab ng-if="$ctrl.isImprovementsTabEnabled()" ng-show="$ctrl.getActiveTabName() === 'improvements'">
```
```suggestion
    <oppia-improvements-tab ng-if="$ctrl.isImprovementsTabEnabled()" 
                            ng-show="$ctrl.getActiveTabName() === 'improvements'">
```

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -1,8 +1,8 @@
     <div class="float-right">
-      <button ng-if="$ctrl.userIsLoggedIn" class="btn btn-success float-right oppia-thread-button oppia-autofocus"
-              ng-click="$ctrl.showCreateThreadModal()" focus-on="newThreadButton">
+      <button *ngIf="userIsLoggedIn" class="btn btn-success float-right oppia-thread-button oppia-autofocus"
+              (click)="showCreateThreadModal()" (focus)="newThreadButton">
```
```suggestion
      <button *ngIf="userIsLoggedIn" 
              class="btn btn-success float-right oppia-thread-button oppia-autofocus"
              (click)="showCreateThreadModal()" 
              (focus)="newThreadButton">
```

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -11,75 +11,77 @@ <h2>Exploration Feedback</h2>
     <div class="oppia-clear"></div>
 
-    <div ng-if="$ctrl.threadData.feedbackThreads.length === 0">
+    <div *ngIf="threadData && threadData.feedbackThreads?.length === 0">
```
Why `?`?

heyimShivam replied:
removed

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -11,75 +11,77 @@ <h2>Exploration Feedback</h2>
       <div class="col-lg-8 col-md-8 col-sm-8">
         <span class="oppia-back-arrow">
-          <button type="button" class="btn btn-secondary btn-sm e2e-test-oppia-feedback-back-button" ng-click="$ctrl.onBackButtonClicked()">
+          <button type="button" class="btn btn-secondary btn-sm e2e-test-oppia-feedback-back-button" (click)="onBackButtonClicked()">
```
```suggestion
          <button type="button" 
                  class="btn btn-secondary btn-sm e2e-test-oppia-feedback-back-button" 
                  (click)="onBackButtonClicked()">
```

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -11,75 +11,77 @@ <h2>Exploration Feedback</h2>
         </span>
-        <span class="oppia-thread-text" ng-show="!$ctrl.activeThread.isSuggestionThread()">Feedback Thread: <[$ctrl.activeThread.subject]></span>
-        <span class="oppia-thread-text" ng-show="$ctrl.activeThread.isSuggestionThread()">Suggestion Thread: <[$ctrl.activeThread.subject]></span>
+        <span class="oppia-thread-text" [hidden]="activeThread.isSuggestionThread()">Feedback Thread: {{ activeThread.subject }}</span>
+        <span class="oppia-thread-text" [hidden]="!activeThread.isSuggestionThread()">Suggestion Thread: {{ activeThread.subject }}</span>
```
```suggestion
        <span class="oppia-thread-text" 
              [hidden]="activeThread.isSuggestionThread()">
          Feedback Thread: {{ activeThread.subject }}
        </span>
        <span class="oppia-thread-text" 
              [hidden]="!activeThread.isSuggestionThread()">
          Suggestion Thread: {{ activeThread.subject }}
        </span>
```

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -11,75 +11,77 @@ <h2>Exploration Feedback</h2>
           </span>
-          <span ng-class="$ctrl.getLabelClass($ctrl.activeThread.status)" class="e2e-test-oppia-feedback-status-name">
-            <[$ctrl.getHumanReadableStatus($ctrl.activeThread.status)]>
+          <span [ngClass]="getLabelClass(activeThread.status)" class="e2e-test-oppia-feedback-status-name">
```
```suggestion
          <span [ngClass]="getLabelClass(activeThread.status)" 
                class="e2e-test-oppia-feedback-status-name">
```

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -11,75 +11,77 @@ <h2>Exploration Feedback</h2>
-                  <div class="oppia-view-suggestion-button-container e2e-test-exploration-feedback"><[m.text]></div>
-                  <div ng-if="$ctrl.activeThread.isSuggestionThread() && $index == 0">
-                    <button class="btn btn-<[$ctrl.getSuggestionButtonType()]> e2e-test-view-suggestion-btn oppia-view-suggestion-btn" ng-click="$ctrl.showSuggestionModal()">
+                  <div class="oppia-view-suggestion-button-container e2e-test-exploration-feedback">{{ m.text }}</div>
```
```suggestion
                  <div class="oppia-view-suggestion-button-container e2e-test-exploration-feedback">
                    {{ m.text }}
                  </div>
```

heyimShivam replied:
done

vojtechjelinek commented on core/templates/pages/exploration-editor-page/feedback-tab/feedback-tab.component.html:
```diff
@@ -11,75 +11,77 @@ <h2>Exploration Feedback</h2>
-                    <button class="btn btn-<[$ctrl.getSuggestionButtonType()]> e2e-test-view-suggestion-btn oppia-view-suggestion-btn" ng-click="$ctrl.showSuggestionModal()">
+                  <div class="oppia-view-suggestion-button-container e2e-test-exploration-feedback">{{ m.text }}</div>
+                  <div *ngIf="activeThread.isSuggestionThread() && index == 0">
+                    <button class="btn btn-{{ getSuggestionButtonType() }} e2e-test-view-suggestion-btn oppia-view-suggestion-btn" (click)="showSuggestionModal()">
```
```suggestion
                    <button class="btn btn-{{ getSuggestionButtonType() }} e2e-test-view-suggestion-btn oppia-view-suggestion-btn" 
                            (click)="showSuggestionModal()">
```

vojtechjelinek replied:
Please try to do similar changes elsewhere, even though we do not follow the 80-line length limit strictly in the HTML files, it is still nice to make the lines shorter.

heyimShivam replied:
okay

vojtechjelinek commented on core/templates/pages/exploration-editor-page/statistics-tab/charts/pie-chart.component.ts:
```diff
@@ -16,74 +16,105 @@
 
-      $scope.$watch('data()', redrawChart);
+   directiveSubscriptions = new Subscription();
+   // TODO(#13015): Remove use of unknown as a type.
```
Remove the TODO.

heyimShivam replied:
removed

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

heyimShivam committed:
html changes

heyimShivam committed:
deleted

heyimShivam committed:
file nits

heyimShivam committed:
white space

heyimShivam committed:
some nits left

heyimShivam commented:
@vojtechjelinek PTAL

heyimShivam committed:
done

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into smallPR1

heyimShivam committed:
kevin changes

heyimShivam commented:
@kevintab95 PTAL

heyimShivam committed:
done

heyimShivam committed:
updated

heyimShivam committed:
fit

heyimShivam committed:
fix

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into smallPR1

heyimShivam committed:
nits

vojtechjelinek reviewed (approved):
LGTM!

kevintab95 reviewed (changes requested):
Thanks for addressing comments @heyimShivam! Looks like there is just one that is not fuly resolved yet: https://github.com/oppia/oppia/pull/16060#discussion_r979450291

Happy to approve once that has been resolved.

heyimShivam commented:
> Thanks for addressing comments @heyimShivam! Looks like there is just one that is not fuly resolved yet: [#16060 (comment)](https://github.com/oppia/oppia/pull/16060#discussion_r979450291)
> 
> Happy to approve once that has been resolved.

i have tried to test these components by enabling improvements tab temporarily what there are console errors, 
can you please share steps to test these components?

kevintab95 commented:
> i have tried to test these components by enabling improvements tab temporarily what there are console errors,

What errors are you getting when you enable the tab? Are there any stack traces printed? Are you able to trace it back to what the root cause is? This is how I would proceed to find out how to test the functionality.

heyimShivam commented:
> > i have tried to test these components by enabling improvements tab temporarily what there are console errors,
> 
> What errors are you getting when you enable the tab? Are there any stack traces printed? Are you able to trace it back to what the root cause is? This is how I would proceed to find out how to test the functionality.


https://user-images.githubusercontent.com/77006897/194411898-51988a2d-c18d-4068-9609-43c6207a0c29.mp4

after sending dummy value to variables this output have been obtained for `oppia-visualization- components`

heyimShivam commented:
@kevintab95 PTAL

heyimShivam committed:
done

heyimShivam commented:
![Screenshot from 2022-10-07 12-44-44](https://user-images.githubusercontent.com/77006897/194493379-8e01ac84-27c0-4d11-a5c3-98b15293b1e8.png)

For Console Error:  the backend provides the image URL, i.e. the image URL  generated in the backend, dummy URL is not working on its place that's why there is console error `Dummy URL used: /assets/images/android/android-landing-speech-bubble.svg`

heyimShivam committed:
updated

heyimShivam committed:
updated

kevintab95 reviewed (approved):
LGTM, thanks @heyimShivam!
